File: packages-gen/dart/substrate/lib/base-pubsub.dart
import 'package:xtalk_substrate/base-router.dart' as router;

import 'package:xtalk_substrate/base-space.dart' as space;

subscribe(node, space, signal, subscription_id, meta) {
  return router.subscribe_frame(space,signal,subscription_id,meta);
}

unsubscribe(node, space, signal, subscription_id, meta) {
  return router.unsubscribe_frame(space,signal,subscription_id,meta);
}

invoke_trigger(node, stream) {
  var signal = stream["signal"];
  var entry = (node["triggers"])[signal];
  if(null == entry){
    return Future.sync(() {
      return null;
    });
  }
  var current_space = space.ensure_space(node,stream["space"],null);
  var trigger_fn = entry["fn"];
  var output = Function.apply(
    (trigger_fn as Function),
    <dynamic>[current_space,stream,node]
  );
  if((() {
    var dart_truthy__41921 = (null != output) && (("Future" == (output.runtimeType).toString()) || (output.runtimeType).toString().startsWith("Future<"));
    return (null != dart_truthy__41921) && (false != dart_truthy__41921);
  })()){
    return output;
  }
  else{
    return Future.sync(() {
      return output;
    });
  }
}

receive_publish(node, stream) {
  space.ensure_space(node,stream["space"],null);
  return ((Future.sync(() => invoke_trigger(node,stream))) as Future<dynamic>).then((value) async { return await Function.apply((_) {
    return stream;
  },<dynamic>[value]); });
}

File: packages-gen/js/libs/xt-substrate/base-pubsub.js
const router = require("@xtalk/substrate/base-router.js")

const space = require("@xtalk/substrate/base-space.js")

function subscribe(node,space,signal,subscription_id,meta){
  return router.subscribe_frame(space,signal,subscription_id,meta);
}

function unsubscribe(node,space,signal,subscription_id,meta){
  return router.unsubscribe_frame(space,signal,subscription_id,meta);
}

function invoke_trigger(node,stream){
  let signal = stream["signal"];
  let entry = (node["triggers"])[signal];
  if(null == entry){
    return Promise.resolve().then(function (){
      return null;
    });
  }
  let current_space = space.ensure_space(node,stream["space"],null);
  let trigger_fn = entry["fn"];
  let output = trigger_fn(current_space,stream,node);
  if(output instanceof Promise){
    return output;
  }
  else{
    return Promise.resolve().then(function (){
      return output;
    });
  }
}

function receive_publish(node,stream){
  space.ensure_space(node,stream["space"],null);
  return invoke_trigger(node,stream).then(function (_){
    return stream;
  });
}

module.exports = {
  ["subscribe"]:subscribe,
  ["unsubscribe"]:unsubscribe,
  ["invoke_trigger"]:invoke_trigger,
  ["receive_publish"]:receive_publish
}